mail_embed_image/models/ir_mail_server.py

```diff
@@ -0,0 +1,130 @@
+            email_cc=None,
+            email_bcc=None,
+            reply_to=None,
+            attachments=None,
+            message_id=None,
+            references=None,
+            object_id=None,
+            subtype='plain',
+            headers=None,
+            body_alternative=None,
+            subtype_alternative='plain'):
+        result = super(IrMailServer, self).build_email(
+            email_from=email_from,
+            email_to=email_to,
+            subject=subject,
+            body=body,
+            email_cc=email_cc,
+            email_bcc=email_bcc,
+            reply_to=reply_to,
+            attachments=attachments,
+            message_id=message_id,
+            references=references,
+            object_id=object_id,
+            subtype=subtype,
+            headers=headers,
+            body_alternative=body_alternative,
+            subtype_alternative=subtype_alternative,
+        )
+        return self._build_email_replace_img_src(result)
+
+    def _build_email_replace_img_src(self, email):
+        """ Given a message, find it's img tags and if they
+        are URLs, replace them with cids.
+        """
+        for part in email.walk():
+            if part.get_content_type() == 'text/html':
+                body = part.get_payload(decode=True)
+                if not body or body == '\n':
+                    continue
+                root = self.process_img_body(fromstring(body), email)
+                # encodestring will put a newline every 74 char
+                part.set_payload(encodestring(tostring(root)))
+        return email
+
+    def process_img_body(self, root, email):
+        base_url = self.env['ir.config_parameter'].get_param(
+            'web.base.url')
+        for img in root.xpath(
+                "//img[starts-with(@src, '%s/web/image')]"
+                "|"
+                "//img[starts-with(@src, '/web/image')]" % (
+                    base_url or '', )):
+            # check if there is a bound request
+            image_path = img.get('src').replace(base_url, '')
+            with self.fetch_image(image_path) as (endpoint, arguments):
+                # now go ahead and call the endpoint and fetch the data
+                response = endpoint.method(**arguments)
+                if not response:
+                    logger.warning('Could not get %s', img.get('src'))
+                    continue
+                cid = uuid.uuid4().hex
+                filename_rfc2047 = encode_header_param(cid)
+                filepart = MIMEImage(response.data)
+                filepart.set_param('name', filename_rfc2047)
+                filepart.add_header(
+                    'Content-Disposition',
+                    'inline',
+                    cid=cid,
+                    filename=filename_rfc2047,
+                )
+                # attach the image into the email as attachment
+                email.attach(filepart)
+                img.set('src', 'cid:%s' % (cid))
```

Comment:
please add a comment here that we should avoid attaching the same image twice. Not necessary for version 1 I think